### ISSUE TITLE
Memory Reservations on Windows Pod

### ISSUE THREAD

**Issue by zhiweiv** [kubernetes/kubernetes#102686]
In https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#memory-reservations,
> As you deploy workloads, use resource limits (must set only limits or _**limits must equal requests**_) on containers.

It says limits must equal requests for memory, does it make sense? We are using following setting and it works.

```
resources:
  requests:
    memory: 1Gi
  limits:
    memory: 4Gi
```

/sig windows

**Comment by immuzz** [comment]
cc : @jsturtevant

**Comment by marosset** [comment]
It does work to set requests lower than limits but in our experience your are likely to run into resource starvation issues on the host when this happens because eviction / memory pressure warnings do not work for windows.

We can update the docs on the website clarify this and say it will work but we strongly recommend against it.

**Comment by zhiweiv** [comment]
Anyway, it is not related to k8s scheduler, right? An Azure support engineer asked me to follow the rule(must set only limits or limits must equal requests) in pod schedule issue, I have doubt with this solution.

**Comment by marosset** [comment]
@zhiweiv can you explain the problems you were seeing that prompted the support engineer to follow this rule?

The centralized scheduler uses requests to decide if the node has enough resources to run the pod. (If requests are omitted then then limits value is used)
The kubelet running on each node is responsible for making sure that the containers in the pods don't use more than the specified resource limits.

On linux it is common practice to set requests lower than limits to get better density assuming that on-average pods are not consuming all the resources up to the limit and are instead using the amount requested. 
Linux has mechanisms to indicate that the node is becoming resource constrained (PID/memory/disk pressure notifications) and also can take actions to kill PODs with the OOM killer.

Windows does not have an OOM killer and under severe memory pressure will not kill pods. 
Instead more virtual memory will be paged to disk for all processes running on the node (including kubelet/container runtime/etc) and you will experience poor performance and/or unresponsive nodes/workloads.

Because of this reason we strongly recommend that memory requests do not be set lower than memory limits unless your environment is finely tuned for your scenarios.

Hope this helps.

**Comment by zhiweiv** [comment]
Thanks for the detailed explanation. 

About the schedule issue, we use bare pods to run background tasks, when task comes, our task controller create a pod to run it, then delete the pod after it run to complete. We found occasionally the scheduler assign all new pods to only one node although there were no pods running in other nodes at same AKS node pool, when the issue happens, it will come back to normal after few hours.

Our task pods have same resource setting as I listed above, so the Azure support engineer ask us to follow the rule to see if it may mitigate schedule issue. 

The schedule issue is still under research, but it should not related to this question based on above explanation. I will keep this open to trace the document update.
